src/arguments.toit

```diff
@@ -39,36 +82,99 @@ class ArgumentParser:
+  $min specifies the minimum number of arguments that must be provided
+    after the options.  It defaults to the length of the $names list, or
+    zero.  When using the length of the $names to determine default
+    $min, trailing arguments that are surrounded by "[]" are not counted, nor
+    are trailing arguments that are named "...".
+  $max specifies the maximum number of arguments that must be provided
+    after the options.  It defaults to the length of the $names list, or
+    $UNLIMITED.  If the last member of the $names list is "..." then there is
+    no maximum by default.
+  $usage is a textual description of the rest arguments that can be
+    provided.  It defaults to a string constructed from the other arguments.
+  */
+  describe_rest -> none
+      names /List? = null
+      --min /int = (calculate_minimum_ names)
+      --max /int = (calculate_maximum_ names)
+      --usage /string? = (calculate_rest_usage_ names min max):
+    if commands_.size != 0: throw "Can't have rest arguments and commands"
+    rest_minimum = min
+    rest_maximum = max
+    rest_usage = usage
+    rest_was_described_ = true
+
+  /**
+  Parses the given $arguments. Returns a new instance of $Arguments.
+  If there is an error it prints the error message and the usage
+    description on stderr, then exits the VM completely with a non-zero
+    exit value.
+  The $invoked_command is used only for the usage message in case of an
+    error.  It defaults to "toit.run $program_name".
+  Returns a new instance of $Arguments.
+  */
+  parse arguments --invoked_command="toit.run $program_name" -> Arguments:
+    return parse arguments --invoked_command=invoked_command: | error_message usage_string |
+      print_on_stderr_ "$error_message"
+      print_on_stderr_
+          usage_string
+      exit 1
+
+  /**
+  Parses the given $arguments.
+  If there is an error in the arguments the block will be called.  If it returns,
+    an exception is thrown.  The arguments to the block are:
+  `error_message`: the error message.
+  `usage_string`: A usage string for the whole parser, or perhaps just the subcommand the user attempted to use.
+    This may be a multiline string, but it doesn't have a terminating newline.
+  The $invoked_command is used only for the usage message in case of an
+    error.  It defaults to "toit.run $program_name".
+  Returns a new instance of $Arguments.
+  */
+  parse arguments --invoked_command="toit.run $program_name" [error_block] -> Arguments:
     try:
       return parse_ this null arguments 0
     finally: | is_exception exception |
       if is_exception:
-        print_on_stderr_ "$exception.value"
-        print_on_stderr_
-            usage arguments
-        if exit_on_error:
-          exit 1
+        (arguments.size + 1).repeat:
+          if it >= arguments.size or not arguments[it].starts_with "-":
+            error_block.call exception.value (usage arguments[it..] --invoked_command=invoked_command)
```

Comment:
I would like to make `finally` not being able to change from exception to `return` or similar.
That said: there are already some uses of this. I guess I will have to revisit this (and discuss it) when I get to it...

https://github.com/toitlang/toit/issues/376